services/QuillLMS/spec/workers/snapshots/cache_snapshot_top_x_worker_spec.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+module Snapshots
+  describe CacheSnapshotTopXWorker, type: :worker do
+    subject { described_class.new }
+
+    let(:cache_key) { 'CACHE_KEY' }
+    let(:query) { 'most-active-schools' }
+    let(:user_id) { 123 }
+    let(:timeframe_name) { 'last-30-days' }
+    let(:school_ids) { [1,2,3] }
+    let(:grades) { ['Kindergarten',1,2,3,4] }
+    let(:query_double) { double }
+
+    context '#perform' do
+      let(:timeframe_end) { DateTime.now }
+      let(:current_timeframe_start) { timeframe_end - 30.days }
+      let(:previous_timeframe_start) { current_timeframe_start - 30.days }
+      let(:timeframe) {
+        {
+          name: timeframe_name,
+          previous_start: previous_timeframe_start,
+          current_start: current_timeframe_start,
+          current_end: timeframe_end
+        }
+      }
+
+      before do
+        stub_const("Snapshots::CacheSnapshotTopXWorker::QUERIES", {
+          query => query_double
+        })
+
+        # Intercept theses calls to isolate them in separate cases
+        allow(query_double).to receive(:run).and_return({})
+        allow(Rails.cache).to receive(:write)
+        allow(PusherTrigger).to receive(:run)
```

Comment:
I think it's better to have specific `expect.to` and `expect.not_to` in each test, rather than a universal allow that some tests overwrite with `expect`. That setup will also catch circumstances when writes or runs are happening and they shouldn't.